### ISSUE TITLE
[BUG] "Cannot open store" when using background `FlutterEngine`

### ISSUE THREAD

**Issue by Jeferson505** [JaffaKetchup/flutter_map_tile_caching#150]
### What is the bug?

I'm developing a navigation app that keeps running even on background mode. I'm using this plugin to download the requested route map on device. The previus plugin's version (8.0.1) was running fine (excpet on Android 9 or old ones, but by another problem that was fixed in this version). However, with the new version (9.0.0) there is a conflict with the background feature.

Basically, without the background feature the FMTCObjectBoxBackend is initialized in main.dart file normally. However, if the application is running in background then when it returns to foreground it generates the follow error:

> StateError (Bad state: failed to create store: Cannot open store: another store is still open using the same path: "/data/data/com.example.app_name/app_flutter/fmtc" (OBX_ERROR code 10001))

That, makes the map screen doesn't be displayed and generates the follow exception:

> RootUnavailable (RootUnavailable: The requested backend/root was unavailable)

### How can we reproduce it?

I created an example project available [here](https://github.com/Jeferson505/fmtc_issue_example)

Observations:
- The example project contains a video and a README file demonstrating how to reproduce the bug;
- I did the native configuration just for Android devices; 
- I wrote all code in a single main.dart file;
- For the project runs as expected is required allow the location permisson in two steps.
	1. While using the app.
	2. All the time.
	* Those permissions that makes possible the application runs on background

The `main.dart` of the example project is:

```dart
import 'dart:developer';

import 'package:background_location_tracker/background_location_tracker.dart';
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:flutter_map_tile_caching/flutter_map_tile_caching.dart';
import 'package:permission_handler/permission_handler.dart';

@pragma('vm:entry-point')
void _backgroundCallback() {
  BackgroundLocationTrackerManager.handleBackgroundUpdated((data) async {
    log('lat: ${data.lat}, long: ${data.lon}');
  });
}

Future<void> _initializeBackgroundLocationTracker() {
  return BackgroundLocationTrackerManager.initialize(
    _backgroundCallback,
    config: const BackgroundLocationTrackerConfig(
      androidConfig: AndroidConfig(
        channelName: 'flutter_background_location_tracker',
        notificationBody: 'Running in the background',
      ),
      loggingEnabled: true,
    ),
  );
}

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await _initializeBackgroundLocationTracker();

  try {
    await FMTCObjectBoxBackend().initialise();
  } catch (e, s) {
    log('Error initialising FMTCObjectBoxBackend', error: e, stackTrace: s);
  }

  runApp(const AppWidget());
}

class AppWidget extends StatelessWidget {
  const AppWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.lightBlue),
      ),
      home: const FlutterMapScreen(),
    );
  }
}

class FlutterMapScreen extends StatefulWidget {
  const FlutterMapScreen({super.key});

  @override
  State<FlutterMapScreen> createState() => _FlutterMapScreenState();
}

class _FlutterMapScreenState extends State<FlutterMapScreen> {
  final _fmtcStore = const FMTCStore('example');
  var _isTracking = false;
  var _isReady = false;

  void _handleIsTracking(bool newValue) {
    if (mounted) setState(() => _isTracking = newValue);
  }

  void _handleIsReady(bool newValue) {
    if (mounted) setState(() => _isReady = newValue);
  }

  void _initIsTracking() {
    BackgroundLocationTrackerManager.isTracking().then(_handleIsTracking);
  }

  void _toggleBackgroundTracking() {
    _isTracking
        ? BackgroundLocationTrackerManager.stopTracking()
        : BackgroundLocationTrackerManager.startTracking();

    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text(
          _isTracking
              ? 'Background location tracking stopped'
              : 'Background location tracking started',
        ),
      ),
    );

    _handleIsTracking(!_isTracking);
  }

  void _requestPermissions() {
    Permission.locationWhenInUse.request().whenComplete(() {
      Permission.locationAlways.request();
    });
  }

  void _tryCreateStore() {
    _fmtcStore.manage.ready.then((exists) {
      if (!exists) {
        _fmtcStore.manage.create().whenComplete(() => _handleIsReady(true));
      } else {
        _handleIsReady(true);
      }
    });
  }

  @override
  void initState() {
    super.initState();
    _initIsTracking();
    _tryCreateStore();
    _requestPermissions();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: FlutterMap(
        options: const MapOptions(),
        children: [
          TileLayer(
            urlTemplate: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
            userAgentPackageName: 'com.example.fmtc_with_background',
            tileProvider: _isReady ? _fmtcStore.getTileProvider() : null,
          ),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _toggleBackgroundTracking,
        child: Icon(
          _isTracking ? Icons.stop_rounded : Icons.play_arrow_rounded,
        ),
      ),
    );
  }
}
```

### Do you have a potential solution?

Before create this issue I did a research to try solve the problem direct in my project code. I identified that's a problem associated with `objectbox` dependencie.

The objectbox-dart issue [436](https://github.com/objectbox/objectbox-dart/issues/436) explain it and [this](https://github.com/objectbox/objectbox-dart/issues/436#issuecomment-1167204502) answer maybe can be a potential solution.

I can't insert that solution direct in my project, because this plugin is the only responsible to use the objectbox dependencie. So, I create this issue to firstly report the problem and after that, try to write a PR to help solve it. But, if someone else has any idea or know a better solution than mine, feel free to share it :)

### Platforms

Android 14 (Pixel 8 Emulator, Xiaomi POCO X5 PRO)

### Severity

Erroneous: Prevents normal functioning and causes errors in the console

**Comment by JaffaKetchup** [comment]
Hey @Jeferson505,
Thanks for reporting this and including so much useful info and research, I definitely would struggle to reproduce this without it!
I'm happy to accept a PR (although please note the dual licensing on this repo means any code you submit will be licensed under GPL, but also sold under alternative licenses without any direct return to you), and I'll also have a look myself and test against the project you provided.

**Comment by Jeferson505** [comment]
" @JaffaKetchup,

Thank you for your response and for all the work you have done on this plugin.

Regarding the license, I am happy to contribute without any financial motivation.

Lastly, I have submitted a [PR](https://github.com/JaffaKetchup/flutter_map_tile_caching/pull/151) with a potential solution. I have tested it and it seems to be working fine. However, for your tests, I recommend considering whether the additional dependency I added is truly necessary. There may be other solutions without the dependency, but I am not aware of them at the moment.

**Comment by JaffaKetchup** [comment]
Thanks for the PR :) I'm a little reluctant to use another dependency, especially one like shared_preferences for what should be a small job, hopefully you understand.

I've done a small amount of playing around with it myself and think I've found a solution that works with your test app in the emulator.

It tests whether the `Store`/root `isOpen` first. If it wasn't, then there's no difference to the current behaviour. Otherwise, we attach it with `Store.attach`.  
There were some issues after that with initialising the thread safe workers, because ObjectBox doesn't set a `ByteData` 'reference' when using `attach` for some reason, which is what they relied on, but using `attach` instead of `fromReference` inside them seems to work exactly the same. I need to determine if there is a difference - the docstring mentions that `fromReference` "guarantees concurrent access", whilst `attach`'s doesn't - and whether it has an effect. I might end up asking ObjectBox's devs why exactly they don't set a reference in this case (although I can see in their internals they have a comment saying they skipped it intentionally because "`attach` is a replacement for `fromReference`"). However, just from some small test it appears to be working at least partially as expected.

I should be able to push a branch with this fix tomorrow, I'd be grateful if you could give it a try at that point, and see if it works as well as your solution :)

**Comment by Jeferson505** [comment]
I really understand you. I completely agree that adding SharedPreferences is not recommended in this scenario.

However, I'm using my fork with that to run my project. So, I shared the solution first of all to show a way to solve it and make it public for anyone else might can face the same problem while we don't find a better solution.

Regarding the use of the `attach` option, before writing the PR solution I tried using it without success. Whenever it throws an exception because the reference was never initialized. After a few unsuccessful attempts, I migrated to `fromReference` implementation.

Perhaps there's an essential modification needed in the ObjectBox plugin. However, I'm available to try any solution you provide.

Once again, thank you so much for all the work and effort you put into this project.

**Comment by leewrigg** [comment]
Hey @JaffaKetchup, Lee from Pitchero here, hope you're well! 👋 

We're also seeing some occurrences of this crash - do you have an ETA for a possible fix?

Thanks

**Comment by JaffaKetchup** [comment]
Hey @leewrigg,

I should be able to release a fix by the end of the day. Been a little busy over the past few days, but there's a fix sitting on my computer. Just need to verify it has no side effects.

**Comment by leewrigg** [comment]
@JaffaKetchup,

Awesome news. I appreciate that you need to have time to verify that the fix works before rolling it out! 

Thank you for the time you put into this plugin, I think it's a testament to the quality of your work that this is the only issue that has been raised after such a large rewrite in v9.

We have only a few instances of this as we don't _generally_ fetch tiles in the background, just in a few situations. 
Can confirm that each time this _did_ occur was while the app was in the background though.

**Comment by JaffaKetchup** [comment]
This should now be fixed with v9.0.1. Thanks for the patience and help!

I couldn't see any functional difference in using `attach` vs `fromReference`, so I've just switched to using `attach`, which has made things slightly simpler. If you notice any difference, please let me know, and I may have to end up asking the ObjectBox team for some insights.
